orthoinference/src/main/java/org/reactome/orthoinference/PathwaysInferrer.java

```diff
@@ -0,0 +1,263 @@
+	private static List<GKInstance> updatedInferrableHumanEvents = new ArrayList<>();
+	private static Map<GKInstance, GKInstance> inferredEventIdenticals = new HashMap<>();
+
+	// This class populates species pathways with the instances that have been inferred. This was copied heavily from the Perl, so my explanations are a little sparse here.
+	public static void inferPathways(List<GKInstance> inferrableHumanEvents) throws Exception
+	{
+		logger.info("Beginning Pathway inference");
+		updatedInferrableHumanEvents.addAll(inferrableHumanEvents);
+
+		// First, go through each of the inferred RlE instances and generate the entire pathway hierarchy it is associated with. Inferred Reactions are not
+		// added to the Pathway at this point. This includes the immediate Pathway, but also all parent pathways up to its TopLevelPathway.
+		logger.info("Building inferred Pathway hierarchies");
+		Set<Long> seenPathwayHierarchy = new HashSet<>();
+		for (GKInstance inferrableInst : inferrableHumanEvents)
+		{
+			logger.info("Building inferred pathways for RlE: " + inferrableInst);
+			if (!seenPathwayHierarchy.contains(inferrableInst.getDBID()))
+			{
+				createInferredPathwayHierarchy(inferrableInst);
+				seenPathwayHierarchy.add(inferrableInst.getDBID());
+			} else {
+				logger.info("Inferred pathways already exist for RlE");
+			}
+		}
+		logger.info("Finished building inferred Pathway hierarchies");
+
+		// After generating the inferred Pathways hierarchys, the associated inferred Events (RlEs & Pathways) need to be added to them.
+		logger.info("Populating inferred Pathways with inferred Events");
+		addInferredEventsToInferredPathways();
+		logger.info("Finished populating inferred Pathways with inferred Events");
+
+		//TODO: LOG starting HERE
+
+		// Connect preceding events to RlEs, if they have any in the source species.
+		logger.info("Adding preceding events to inferred Events");
+		inferPrecedingEvents();
+		logger.info("Finished adding preceding events to inferred Events");
+
+		// Any source species Events (Pathway or RlEs) that were modified during Pathway inference are updated with a 'modified' instance edit.
+		updateModifiedAttributeIfNecessary();
+	}
+
+	@SuppressWarnings("unchecked")
+	// This generates the inferred Pathway of an inferred RlE. It iterates, inferring parent Pathways until reaching the TopLevelPathway.
+	// Inferred Reactions are not added to the Pathway at this step.
+	private static void createInferredPathwayHierarchy(GKInstance sourceEventInst) throws Exception
+	{
+		List<GKInstance> sourcePathwayReferralInstances = (ArrayList<GKInstance>) sourceEventInst.getReferers(hasEvent);
+
+		if (sourcePathwayReferralInstances != null && sourcePathwayReferralInstances.size() > 0)
+		{
+			for (GKInstance sourcePathwayReferralInst : sourcePathwayReferralInstances)
+			{
+				logger.info("Generating inferred Pathway: " + sourcePathwayReferralInst);
+				if (inferredEventIdenticals.get(sourcePathwayReferralInst) == null)
```

Comment:
Could this logic be extracted to a method called something like `inferPathway`?  This would make this method easier to read and allow for unit testing of inferring a single pathway.